src/spikeinterface/generation/hybrid_tools.py

```diff
@@ -0,0 +1,562 @@
+        generate_sorting_kwargs["seed"] = seed
+        sorting = generate_sorting(**generate_sorting_kwargs)
+    else:
+        assert sorting.sampling_frequency == sampling_frequency
+
+    num_spikes = sorting.to_spike_vector().size
+    sorting.set_property("gt_unit_locations", unit_locations)
+
+    assert (nbefore + nafter) == templates_array.shape[
+        1
+    ], "templates and ms_before, ms_after should have the same length"
+
+    if templates_array.ndim == 3:
+        upsample_vector = None
+    else:
+        if upsample_vector is None:
+            upsample_factor = templates_array.shape[3]
+            upsample_vector = rng.integers(0, upsample_factor, size=num_spikes)
+
+    if amplitude_std is not None:
+        amplitude_factor = rng.normal(loc=1, scale=amplitude_std, size=num_spikes)
+    else:
+        amplitude_factor = None
+
+    if motion is not None:
+        assert num_segments == motion.num_segments, "recording and motion should have the same number of segments"
+        dim = motion.dim
+        motion_array_concat = np.concatenate(motion.displacement)
+        if dim == 0:
+            start = np.array([np.min(motion_array_concat), 0])
+            stop = np.array([np.max(motion_array_concat), 0])
+        elif dim == 1:
+            start = np.array([0, np.min(motion_array_concat)])
+            stop = np.array([0, np.max(motion_array_concat)])
+        elif dim == 2:
+            raise NotImplementedError("3D motion not implemented yet")
+        displacements = make_linear_displacement(start, stop, num_step=int((stop - start)[dim]))
+
+        # use templates_, because templates_array might have been scaled
+        drifting_templates = DriftingTemplates.from_static_templates(templates_)
+        drifting_templates.precompute_displacements(displacements)
+
+        # calculate displacement vectors for each segment
+        displacement_sampling_frequency = 1.0 / np.diff(motion.temporal_bins_s[0])[0]
+        displacement_vectors = []
+        spatial_bins_um = motion.spatial_bins_um
+        for segment_index in range(motion.num_segments):
+            temporal_bins_segment = motion.temporal_bins_s[segment_index]
+            displacement_segment = motion.displacement[segment_index]
+            displacement_vector = np.zeros((len(temporal_bins_segment), 2, len(spatial_bins_um)))
+            displacement_unit_factor = np.zeros((num_units, len(spatial_bins_um)))
+
+            for count, i in enumerate(spatial_bins_um):
+                local_motion = displacement_segment[:, count]
+                displacement_vector[:, motion.dim, count] = local_motion
+            displacement_vectors.append(displacement_vector)
+
+        # calculate displacement unit factor
+        displacement_unit_factor = np.zeros((num_units, len(spatial_bins_um)))
+        for count in range(num_units):
+            a = 1 / np.abs((unit_locations[count, motion.dim] - spatial_bins_um))
```

Comment:
I'm a bit confused about this -- why not just have each unit follow the motion at its location? (I.e., compute one displacement vector for each unit and `displacement_unit_factor = np.eye(num_units)`.) This seems less intuitive and more like a specific choice (why 1/r weighting?)

Reply:
that's a good point @cwindolf !!!

Reply:
@samuelgarcia @yger can you comment on this?

Reply:
@cwindolf in pratice we don't have the motion at the unit depth, so we need to interpolate/weigh the estimated motions depending on the unit location. This is one way to do that, so that if you have a motion estimated at a certain depth it will be predominant (1/r).

Reply:
@cwindolf : I think this could be done but the displacement vector will be enormous and I think this would slow down the generator. Here alessio is implementing one displacement per spatial bin.
@alejoe91 : I would use a linear interpolation between the 2 nearest spatial bins no ?

Reply:
I think @yger did this

Reply:
Yeah, the `Motion` object knows how to do interpolation already, and it uses linear interpolation rather than this distance-weighted one -- it could be nice to use the Motion object internally, but it's nice to be able to control drift per-unit like this allows. Maybe a middle ground could be to generate a linear interpolation matrix here, so that the output matches what you would get when using the motion object?

Reply:
Done in the last commits @cwindolf !!!